### PR TITLE
Add function to improve nested dict config parsing

### PR DESCRIPTION
**Pull request by davebulaval**
Hydra-core uses a nested dict config format to represent config. Such data structure is poorly parsed actually (see picture).
Screenshot from 2022-01-13 15-57-28

Do
- This PR improves the parsing of nested dict structure (DictConfig) along with sequence structure. It results in more readable config logging (see picture below). 
- This PR also fixes the inconsistent doc-hint typing for `init` config parameter.
- This PR introduces base cases tests for `parse_config`.
- This PR introduces tests for newly introduce parsing.

Don't
-  I've tried adding hint typing to parse_config but `mypy` does not like the missing FLAGS type missing from Tensorflow. 
- If PR is accepted, will improve related doc to config logging.

Screenshot from 2022-01-13 16-05-24

### CODE REVIEW

File: wandb/sdk/wandb_helper.py [hunk 1]
```diff
@@ -47,12 +61,63 @@ def _to_dict(params):
...
+    if isinstance(element, Mapping):
+        # Case where the value is another dict (a nested dict)
+        unwrapped_nested_params = {}
+        for key, value in element.items():
+            # We recursively open the element (Dict format type)
+            unwrapped_nested_params.update(
+                _unwrap_nested_config(f"{parent_name}.{key}", value)
+            )
+        return unwrapped_nested_params
+    elif isinstance(element, Sequence) and not isinstance(element, str):
```
Comment (dmitryduev) [inline wandb/sdk/wandb_helper.py:113]:
```suggestion
    elif isinstance(element, Sequence) and not isinstance(element, (str, bytes)):
```

**Comment by davebulaval** [comment]
@dmitryduev, do you have an ETA to the merge of my PR?

**Comment by dmitryduev** [comment]
hey @davebulaval, many thanks for the excellent contribution and sorry about the slow response!
we have been thinking about how to best incorporate this as it is really nice to be able to natively support hydra/nested mappings in general. our main concern is with the way the (nested) config keys are flattened, which makes the transformation non-invertible. 
we are currently hacking on a frontend component that will allow (interactively) displaying nested configs on the run info page. how does that sound to you?

**Comment by davebulaval** [comment]
@dmitryduev, I agree that flatten config is less readable in the GUI. I agree that hacking to show the config in a nested way is way more interesting. Keep me updated!

**Comment by dmitryduev** [comment]
Our current thinking is that it'll look something like what we have in the [run comparer](https://docs.wandb.ai/ref/app/features/panels/run-comparer); so we'd want to merge in most of what you have in this pr, except for the nested key flattening logic.
Thanks again, and I'll keep you posted!

**Comment by davebulaval** [comment]
@dmitryduev, it seems like it.

**Comment by dmitryduev** [comment]
@davebulaval a quick update: we have the frontend PR in review now - I'll ping you once that is merged.

**Comment by dmitryduev** [comment]
hey @davebulaval: our frontend can now display nested mappings:
image
Please give it a try! 
Would you be willing to repurpose this PR into one that adds native support of Mapping objects like OmegaConf's dictionary (using e.g. `omegaconf.OmegaConf.to_container`), i.e. so that one could feed those directly into `wandb.init(config=...)`?

**Comment by davebulaval** [comment]
@dmitryduev, to be sure, you want to add a clear implementation in `parse_config` that will use `OmegaConf.to_container(conf, resolve=True)` to handle the conversion of OmegaConf into primitive dict?
Do I keep the other nested approach? Or we will only focus on OmegaConf-like nested dict?
I think that Hydra-core uses OmegaConf, but I haven't validated it.

**Comment by davebulaval** [comment]
> @dmitryduev, to be sure, you want to add a clear implementation in `parse_config` that will use `OmegaConf.to_container(conf, resolve=True)` to handle the conversion of OmegaConf into primitive dict? Do I keep the other nested approach? Or we will only focus on OmegaConf-like nested dict? I think that Hydra-core uses OmegaConf, but I haven't validated it.

**Comment by davebulaval** [comment]
My bad. @dmitryduev I'm still waiting for responses to my previous questions.

**Comment by dmitryduev** [comment]
Hey @davebulaval, apologies for the late reply.

> add a clear implementation in `parse_config` that will use `OmegaConf.to_container(conf, resolve=True)` to handle the conversion of OmegaConf into primitive dict? 
Yes, please!
> Do I keep the other nested approach?
Basically, we only need to make sure that the config is a (true) nested dict, so I'd only focus on OmegaConf for now (with a test case that a hydra config gets processed properly).

Thanks once again for your contribution!

**Comment by davebulaval** [comment]
@dmitryduev sorry for the long delay, I've been busy. I've just pushed a fix with what you are asking.

**Comment by davebulaval** [comment]
I've just pushed a fix for the missing requirements `omegaconf` in the requirements. I don't know If there is another place that is indicated for the doc.

Also, flake and `mypy` is not passing due to import order and frankly, I don't know what to do with it. Flake comply about placing order and I'm not familiar with that. I personally prefer Black.

**Pull request closed without merging**